File: platform/macos/Storage/NativeCertificateTrust.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Combine
import Foundation

@MainActor public protocol NativeTrustTarget: AnyObject {
  var prompt: NativePrompt? { get }
  var generation: UInt64 { get }
  var isClosing: Bool { get }
  func replyTrust(to request: NativePrompt, allowed: Bool) throws
}
extension NativeSession: NativeTrustTarget {}

@MainActor public final class NativeCertificateTrust: ObservableObject {
  @Published public private(set) var inspection: NativeLegacyTrustMatch?
  @Published public private(set) var savedInspection: NativeSavedTrustInspection?
  @Published public private(set) var savedIssue: NativeStorageError?
  @Published public private(set) var notice: String?
  @Published public private(set) var needsReload = false
  @Published public private(set) var issue: NativeTrustStoreIssue?
  @Published public private(set) var isWorking = false
  private weak var target: (any NativeTrustTarget)?
  private let store: NativeLegacyTrustStore?
  private let savedStore: NativeTrustStore?
  private var scope: NativeTrustScope?
  private var hostScope: NativeTrustScope?
  private let hostKeyStore: NativeTrustStore?
  private var latest: NativePrompt?
  private var work: Task<Void,Never>?
  private var epoch: UInt64 = 0
  private var suspended = false, stopped = false
  public init(store: NativeLegacyTrustStore? = nil, savedStore: NativeTrustStore? = nil, hostKeyStore: NativeTrustStore? = nil) { self.store = store; self.savedStore = savedStore; self.hostKeyStore = hostKeyStore }
  public func bind(_ target: any NativeTrustTarget) { self.target = target }
  public func beginAttempt(endpoint: String? = nil, routeIdentity: String = "") {
    suspended = false; scope = endpoint.flatMap { try? NativeTrustScope(endpoint: $0,routeIdentity: routeIdentity) }
    hostScope = endpoint.flatMap { try? NativeTrustScope(endpoint: $0,routeIdentity: routeIdentity,kind: .hostKey) }; inspect(nil)
  }
  public func inspect(_ request: NativePrompt?) {
    epoch &+= 1; work?.cancel(); latest = request; inspection = nil; issue = nil; savedInspection = nil; savedIssue = nil; notice = nil; needsReload = false
    if work == nil { startIfNeeded() }
  }
  private func current(_ request: NativePrompt, ticket: UInt64) -> Bool {
    !stopped && !suspended && epoch == ticket && !Task.isCancelled &&
      target?.isClosing == false && target?.generation == request.generation && target?.prompt == request
  }
  private func startIfNeeded() {
    guard !stopped, !suspended, let request = latest, request.kind == .certificate || request.kind == .hostKey,
          NativeTrustPresentation(request).mayConnectOnce else { return }
    let store = store, savedStore = request.kind == .hostKey ? hostKeyStore : savedStore, scope = request.kind == .hostKey ? hostScope : scope
    let ticket = epoch; isWorking = true
    work = Task { [weak self] in
      defer { self?.finish(ticket: ticket) }
      do {
        if let savedStore {
          guard let scope else { throw NativeStorageError.invalid }
          let saved: NativeSavedTrustInspection
          if request.kind == .hostKey { saved = try await savedStore.inspectHostKey(scope: scope,key: request.identity) }
          else { saved = try await savedStore.inspect(scope: scope,certificate: request.identity) }
          guard let self, self.current(request,ticket: ticket) else { return }
          self.savedInspection = saved
          if saved.state == .match { try self.target?.replyTrust(to: request,allowed: true); return }
          // Explicit endpoint decisions take precedence. Forgetting never revives
          // a pre-existing broad legacy exception for this destination.
          if saved.state != .absent { return }
        }
        if request.kind == .hostKey { return }
        guard let store else { throw NativeTrustStoreIssue.unavailable }
        let result = try await store.lookup(host: request.serverName, certificate: request.identity)
        guard let self, self.current(request,ticket: ticket) else { return }
        self.inspection = result
        if result.state == .match {
          // Preserve the legacy exception decision only for this currently
          // pending certificate and after the non-overridable policy gate.
          try self.target?.replyTrust(to: request, allowed: true)
        }
      } catch {
        guard let self, self.current(request,ticket: ticket) else { return }
        if let saved = error as? NativeStorageError { self.savedIssue = saved; self.needsReload = true }
        else { self.issue = error as? NativeTrustStoreIssue ?? .unavailable }
      }
    }
  }
  private func finish(ticket: UInt64) {
    work = nil; isWorking = false
    if epoch != ticket { startIfNeeded() }
  }
  public func canConnectOnce(_ request: NativePrompt) -> Bool {
    !isWorking && current(request,ticket: epoch) && NativeTrustPresentation(request).mayConnectOnce
  }
  public func connectOnce(_ request: NativePrompt) throws {
    guard canConnectOnce(request), let target else { throw NativeError(.stale,"Inactive trust request") }
    try target.replyTrust(to: request, allowed: true)
  }
  public func canSave(_ request: NativePrompt) -> Bool {
    canConnectOnce(request) && (request.kind == .hostKey ? hostKeyStore != nil && hostScope != nil : savedStore != nil && scope != nil) &&
      savedInspection != nil && savedInspection?.state != .match && savedIssue == nil && !needsReload
  }
  public var replacesSavedKey: Bool { savedInspection?.state == .changed }
  public func reload() { if work == nil, let latest { inspect(latest) } }
  public func saveAndConnect(_ request: NativePrompt) { change(request,forget: false) }
  public func forget(_ request: NativePrompt) { change(request,forget: true) }
  private func change(_ request: NativePrompt, forget: Bool) {
    guard canSave(request), let savedStore = request.kind == .hostKey ? hostKeyStore : savedStore,
          let scope = request.kind == .hostKey ? hostScope : scope, let saved = savedInspection else { return }
    let ticket = epoch; isWorking = true; notice = nil; savedIssue = nil
    work = Task { [weak self] in
      defer { self?.finish(ticket: ticket) }
      do {
        let result: NativeTrustCommit
        if forget { result = try await savedStore.forget(scope: scope,expected: saved.revision) }
        else if request.kind == .hostKey { result = try await savedStore.saveHostKey(scope: scope,key: request.identity,replacing: saved.state == .changed,expected: saved.revision) }
        else { result = try await savedStore.save(scope: scope,certificate: request.identity,status: request.certificateStatus,
          replacing: saved.state == .changed,expected: saved.revision) }
        guard let self, self.current(request,ticket: ticket) else { return }
        self.inspection = nil
        self.savedInspection = .init(state: forget ? .forgotten : .match,revision: result.snapshot.revision,
          expectedFingerprint: forget ? nil : saved.receivedFingerprint,receivedFingerprint: saved.receivedFingerprint)
        if result.durabilityUncertain {
          self.needsReload = true
          self.notice = "The decision was written, but its durability could not be confirmed. Reload to check the saved decision."
        } else if forget {
          self.notice = request.kind == .hostKey ? "Forgot the saved server key for this destination. Compare the key again before continuing." : "Forgot the saved key for this destination. Older host-wide exceptions will not be reused here."
        } else { try self.target?.replyTrust(to: request,allowed: true) }
      } catch {
        guard let self, self.current(request,ticket: ticket) else { return }
        self.savedIssue = error as? NativeStorageError ?? .ioFailure; self.needsReload = true
      }
    }
  }
  public func cancel() { suspended = true; inspect(nil) }
  public func stop() { stopped = true; cancel(); target = nil }
  public func close() async { stop(); await work?.value }
  public var issueMessage: String? {
    if let savedIssue { return nativeTrustStorageMessage(savedIssue) }
    guard let issue else { return nil }
    switch issue {
    case .denied: return "Saved certificate exceptions could not be read because access was denied."
    case .unsafeFile: return "The saved certificate exception file has unsafe ownership, permissions or file type."
    case .corrupt: return "The saved certificate exception file is malformed. It has not been changed."
    case .unsupportedFormat, .unsupportedDigest: return "The saved certificate exceptions use an unsupported format or digest."
    case .tooLarge: return "The saved certificate exception file exceeds the supported size."
    case .changed: return "The saved certificate exceptions changed while being read. Connect again to check them."
    case .cancelled, .closed: return "The certificate exception check was cancelled."
    case .unavailable: return "Saved certificate exceptions could not be checked."
    }
  }
}

File: platform/windows/TidyVNC.Native/Trust/NativeCertificateTrust.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using System.ComponentModel;
using CommunityToolkit.Mvvm.ComponentModel;
using TidyVNC.Native.Storage;

namespace TidyVNC.Native.Trust;

public enum NativeTrustNotice { Forgotten, ForgottenHostKey }

/// <summary>
/// One per connection window (NativeCertificateTrust on macOS; TRUST.md,
/// SERVICES.md section 5). For a certificate or server-key prompt the core
/// may let the user override, it checks the saved decision for this
/// destination first: a match answers the prompt; a forgotten or changed
/// entry stops there (a forget never revives a broader legacy exception).
/// With no saved entry, a certificate is looked up in the legacy files and a
/// match answers the prompt. Otherwise the dialog offers Connect once, a
/// confirmed Save exception and connect, and Cancel. Verification itself
/// stays in the core. UI thread only.
/// </summary>
public sealed partial class NativeCertificateTrust : ObservableObject
{
    private readonly IUiDispatcher dispatcher;
    private readonly NativeLegacyTrustFiles? legacy;
    private readonly NativeTrustStore? certificates, hostKeys;
    private NativeSession? target;
    private NativeTrustScope? scope, hostScope;
    private NativePrompt? latest;
    private Task? work;
    private CancellationTokenSource? workCancel;
    private ulong epoch;
    private bool suspended, stopped;

    [ObservableProperty] public partial NativeKnownHostsMatch? Inspection { get; private set; }
    [ObservableProperty] public partial NativeSavedTrustInspection? SavedInspection { get; private set; }
    [ObservableProperty] public partial NativeStorageError? SavedIssue { get; private set; }
    [ObservableProperty] public partial NativeLegacyTrustError? Issue { get; private set; }
    [ObservableProperty] public partial NativeTrustNotice? Notice { get; private set; }
    [ObservableProperty] public partial bool NeedsReload { get; private set; }
    [ObservableProperty] public partial bool IsWorking { get; private set; }

    public NativeCertificateTrust(IUiDispatcher dispatcher, NativeLegacyTrustFiles? legacy = null, NativeTrustStore? certificates = null,
                                  NativeTrustStore? hostKeys = null)
    {
        if (certificates is { Kind: not NativeTrustKind.Certificate } || hostKeys is { Kind: not NativeTrustKind.HostKey })
            throw new ArgumentException("Trust store kinds do not match");
        this.dispatcher = dispatcher;
        this.legacy = legacy;
        this.certificates = certificates;
        this.hostKeys = hostKeys;
    }

    public Task Work => work ?? Task.CompletedTask;

    public void Bind(NativeSession session)
    {
        UiThread.Require(dispatcher);
        if (target is { } previous) previous.PropertyChanged -= OnSessionChanged;
        target = session;
        session.PropertyChanged += OnSessionChanged;
    }

    private void OnSessionChanged(object? sender, PropertyChangedEventArgs change)
    {
        if (ReferenceEquals(sender, target) && change.PropertyName == nameof(NativeSession.Prompt)) Inspect(target!.Prompt);
    }

    public void BeginAttempt(string? endpoint = null, string routeIdentity = "")
    {
        UiThread.Require(dispatcher);
        suspended = false;
        scope = hostScope = null;
        if (endpoint is { Length: > 0 })
        {
            try
            {
                scope = NativeTrustScope.Create(endpoint, routeIdentity, NativeTrustKind.Certificate);
                hostScope = NativeTrustScope.Create(endpoint, routeIdentity, NativeTrustKind.HostKey);
            }
            catch (NativeIdentityFailure) { scope = hostScope = null; }
        }
        Inspect(null);
    }

    private void Inspect(NativePrompt? request)
    {
        ++epoch;
        workCancel?.Cancel();
        latest = request;
        Inspection = null; Issue = null; SavedInspection = null; SavedIssue = null; Notice = null; NeedsReload = false;
        if (work is null) StartIfNeeded();
    }

    private bool Current(NativePrompt request, ulong ticket, CancellationToken token)
        => !stopped && !suspended && epoch == ticket && !token.IsCancellationRequested && target is { IsClosing: false } session &&
           session.Generation == request.Generation && session.Prompt is { } prompt && prompt.Id == request.Id && prompt.Generation == request.Generation;

    private NativeTrustStore? StoreFor(NativePrompt request) => request.Kind == NativePrompt.PromptKind.HostKey ? hostKeys : certificates;
    private NativeTrustScope? ScopeFor(NativePrompt request) => request.Kind == NativePrompt.PromptKind.HostKey ? hostScope : scope;

    private void StartIfNeeded()
    {
        if (stopped || suspended || latest is not { } request ||
            request.Kind is not (NativePrompt.PromptKind.Certificate or NativePrompt.PromptKind.HostKey) ||
            !new NativeTrustPresentation(request).MayConnectOnce) return;
        var saved = StoreFor(request);
        var destination = ScopeFor(request);
        Start(async (ticket, token) =>
        {
            try
            {
                if (saved is not null)
                {
                    if (destination is null) throw new NativeStorageException(NativeStorageError.Unavailable);
                    var inspection = request.Kind == NativePrompt.PromptKind.HostKey
                        ? await saved.InspectHostKeyAsync(destination, request.Identity, token)
                        : await saved.InspectAsync(destination, request.Identity, token);
                    if (!Current(request, ticket, token)) return;
                    SavedInspection = inspection;
                    if (inspection.State == NativeSavedTrustState.Match) { target!.ReplyTrust(request, true); return; }
                    // Explicit destination decisions win; a forget never revives a legacy exception.
                    if (inspection.State != NativeSavedTrustState.Absent) return;
                }
                if (request.Kind == NativePrompt.PromptKind.HostKey) return;
                if (legacy is null) throw new NativeLegacyTrustException(NativeLegacyTrustError.Unavailable);
                var result = await legacy.LookupAsync(request.ServerName, request.Identity, token);
                if (!Current(request, ticket, token)) return;
                Inspection = result;
                // Only for this pending certificate, after the core's non-overridable policy gate.
                if (result.State == NativeKnownHostsState.Match) target!.ReplyTrust(request, true);
            }
            catch (Exception error) when (error is NativeStorageException or NativeLegacyTrustException or NativeError)
            {
                if (!Current(request, ticket, token)) return;
                if (error is NativeStorageException storage) { SavedIssue = storage.Error; NeedsReload = true; }
                else Issue = (error as NativeLegacyTrustException)?.Error ?? NativeLegacyTrustError.Unavailable;
            }
        });
    }

    private void Start(Func<ulong, CancellationToken, Task> body)
    {
        var ticket = epoch;
        var cancel = new CancellationTokenSource();
        workCancel = cancel;
        IsWorking = true;
        work = Run();

        async Task Run()
        {
            try
            {
                await Task.Yield();
                await body(ticket, cancel.Token);
            }
            finally
            {
                cancel.Dispose();
                if (ReferenceEquals(workCancel, cancel)) workCancel = null;
                work = null;
                IsWorking = false;
                if (epoch != ticket) StartIfNeeded();
            }
        }
    }

    public bool CanConnectOnce(NativePrompt request)
        => !IsWorking && Current(request, epoch, CancellationToken.None) && new NativeTrustPresentation(request).MayConnectOnce;

    /// <summary>The dialog's Connect once: this attempt only, nothing saved.</summary>
    public void ConnectOnce(NativePrompt request)
    {
        UiThread.Require(dispatcher);
        if (!CanConnectOnce(request) || target is null) throw new NativeError(NativeStatus.Stale, "Inactive trust request");
        target.ReplyTrust(request, true);
    }

    public bool CanSave(NativePrompt request)
        => CanConnectOnce(request) && StoreFor(request) is not null && ScopeFor(request) is not null &&
           SavedInspection is { State: not NativeSavedTrustState.Match } && SavedIssue is null && !NeedsReload;

    /// <summary>True when saving replaces a different saved key (the confirmation says so).</summary>
    public bool ReplacesSavedKey => SavedInspection?.State == NativeSavedTrustState.Changed;

    public void Reload()
    {
        UiThread.Require(dispatcher);
        if (work is null && latest is { } request) Inspect(request);
    }

    /// <summary>The confirmed Save exception and connect.</summary>
    public void SaveAndConnect(NativePrompt request) => Change(request, forget: false);

    public void Forget(NativePrompt request) => Change(request, forget: true);

    private void Change(NativePrompt request, bool forget)
    {
        UiThread.Require(dispatcher);
        if (!CanSave(request) || StoreFor(request) is not { } saved || ScopeFor(request) is not { } destination || SavedInspection is not { } inspection) return;
        Notice = null;
        SavedIssue = null;
        Start(async (ticket, token) =>
        {
            try
            {
                var result = forget ? await saved.ForgetAsync(destination, inspection.Revision, CancellationToken.None)
                    : request.Kind == NativePrompt.PromptKind.HostKey
                        ? await saved.SaveHostKeyAsync(destination, request.Identity, inspection.State == NativeSavedTrustState.Changed, inspection.Revision, CancellationToken.None)
                        : await saved.SaveAsync(destination, request.Identity, request.CertificateStatus, inspection.State == NativeSavedTrustState.Changed,
                                                inspection.Revision, CancellationToken.None);
                if (!Current(request, ticket, token)) return;
                Inspection = null;
                SavedInspection = new(forget ? NativeSavedTrustState.Forgotten : NativeSavedTrustState.Match, result.Revision,
                    forget ? null : inspection.ReceivedFingerprint, inspection.ReceivedFingerprint);
                if (forget) Notice = request.Kind == NativePrompt.PromptKind.HostKey ? NativeTrustNotice.ForgottenHostKey : NativeTrustNotice.Forgotten;
                else target!.ReplyTrust(request, true);
            }
            catch (Exception error) when (error is NativeStorageException or NativeError)
            {
                if (!Current(request, ticket, token)) return;
                SavedIssue = (error as NativeStorageException)?.Error ?? NativeStorageError.IOFailure;
                NeedsReload = true;
            }
        });
    }

    /// <summary>The dialog's Cancel (the default button): stops inspecting until the next attempt.</summary>
    public void Cancel()
    {
        UiThread.Require(dispatcher);
        suspended = true;
        Inspect(null);
    }

    public void Stop()
    {
        UiThread.Require(dispatcher);
        stopped = true;
        Cancel();
        if (target is not null) target.PropertyChanged -= OnSessionChanged;
        target = null;
    }

    public async Task CloseAsync()
    {
        Stop();
        await Work;
    }
}

/// <summary>A trust library window's model (NativeTrustLibrary on macOS): list, reload and forget.</summary>
public sealed partial class NativeTrustLibrary(IUiDispatcher dispatcher, NativeTrustStore store) : ObservableObject
{
    private Task? work;
    private bool stopped;

    [ObservableProperty] public partial NativeSavedTrustSnapshot? Snapshot { get; private set; }
    [ObservableProperty] public partial bool IsWorking { get; private set; }
    [ObservableProperty] public partial bool NeedsReload { get; private set; }
    [ObservableProperty] public partial NativeStorageError? Issue { get; private set; }
    [ObservableProperty] public partial bool Forgot { get; private set; }

    public NativeTrustKind Kind => store.Kind;
    public IReadOnlyList<NativeSavedTrustEntry> Entries => Snapshot?.Entries ?? [];
    public Task Work => work ?? Task.CompletedTask;

    private void Start(Func<Task> body)
    {
        IsWorking = true;
        Issue = null;
        Forgot = false;
        work = Run();

        async Task Run()
        {
            try
            {
                await Task.Yield();
                await body();
            }
            catch (NativeStorageException error)
            {
                if (!stopped) { NeedsReload = true; Issue = error.Error; }
            }
            finally
            {
                work = null;
                IsWorking = false;
            }
        }
    }

    public void Reload()
    {
        UiThread.Require(dispatcher);
        if (stopped || work is not null) return;
        Start(async () =>
        {
            var snapshot = await store.ReadSnapshotAsync();
            if (stopped) return;
            Snapshot = snapshot;
            OnPropertyChanged(nameof(Entries));
            NeedsReload = false;
        });
    }

    public void Forget(string id)
    {
        UiThread.Require(dispatcher);
        if (stopped || work is not null || NeedsReload || Snapshot is not { } snapshot ||
            snapshot.Entries.FirstOrDefault(e => e.Id == id) is not { IsForgotten: false } entry) return;
        Forget(entry.Scope, snapshot.Revision);
    }

    public void ForgetDestination(string endpoint, string routeIdentity = "")
    {
        UiThread.Require(dispatcher);
        if (stopped || work is not null || NeedsReload || Snapshot is not { } snapshot) return;
        NativeTrustScope scope;
        try { scope = NativeTrustScope.Create(endpoint, routeIdentity, store.Kind); }
        catch (NativeIdentityFailure) { Issue = NativeStorageError.Corrupt; return; }
        Forget(scope, snapshot.Revision);
    }

    private void Forget(NativeTrustScope scope, Guid? revision) => Start(async () =>
    {
        var result = await store.ForgetAsync(scope, revision);
        if (stopped) return;
        Snapshot = result;
        OnPropertyChanged(nameof(Entries));
        Forgot = true;
    });

    public async Task CloseAsync()
    {
        stopped = true;
        await Work;
    }
}
